internal/handlers/sql/msg_count.go

```diff
@@ -0,0 +1,106 @@
+// Copyright 2021 FerretDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/jackc/pgx/v4"
+
+	"github.com/FerretDB/FerretDB/internal/handlers/common"
+	"github.com/FerretDB/FerretDB/internal/pg"
+	"github.com/FerretDB/FerretDB/internal/types"
+	"github.com/FerretDB/FerretDB/internal/util/lazyerrors"
+	"github.com/FerretDB/FerretDB/internal/wire"
+)
+
+// MsgCount counts the number of documents that matches the query filter.
+func (s *storage) MsgCount(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, error) {
+	// TODO cursor / getMore support via https://www.postgresql.org/docs/current/sql-declare.html
+
+	document, err := msg.Document()
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	if err := common.Unimplemented(document, common.UnimplementedFieldsFindOrCount...); err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+	common.Ignored(document, s.l.Desugar(), common.IgnoredFieldsFindOrCount...)
+
+	var filter *types.Document
+	var sql, collection string
+
+	m := document.Map()
+	db := m["$db"].(string)
+
+	collection = m["count"].(string)
+	filter, _ = m["query"].(*types.Document)
+	sql = fmt.Sprintf(`SELECT COUNT(*) FROM %s`, pgx.Identifier{db, collection}.Sanitize())
+
+	var placeholder pg.Placeholder
+
+	whereSQL, args, err := where(filter, &placeholder)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	sql += whereSQL
+	sort, _ := m["sort"].(*types.Document)
+	limit, _ := m["limit"].(int32)
+
+	tailSQL, tailArgs, err := common.Tail(sort, limit, " ", &placeholder)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+	args = append(args, tailArgs...)
+	sql += tailSQL
+
+	rows, err := s.pgPool.Query(ctx, sql, args...)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+	defer rows.Close()
+
+	var res wire.OpMsg
+
+	var count int32
+	for rows.Next() {
+		err := rows.Scan(&count)
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+	}
+	// in psql, the SELECT * FROM table limit `x` ignores the value of the limit,
```

Comment:
psql should not ignore LIMIT in `SELECT * FROM table LIMIT x`. WDYM?

https://github.com/FerretDB/FerretDB/pull/331#discussion_r818826776